### PR TITLE
feat(spec): SpecInput data contract + host/device boundary annotations

### PR DESCRIPTION
**Pull request by zorrofox**
Part of #1053 (P1-5a). **Draft for early interface review** — P1-2 (`BaseSpecWorker`/`BaseDraftWorker` class refactor) builds on this contract, so opening early per [discussion](https://github.com/sgl-project/sglang-jax/issues/1053#issuecomment-4427601694).

> [!NOTE]
> **Depends on #1063** (P1-0). This branch is stacked on it; the actual P1-5a diff is the last commit only (4 files, +201/−26). Will rebase onto main once #1063 merges.

## What

Defines the spec-decode data contract from the #1053 RFC tables so the P1-2 abstraction layer has a fixed interface to target:

- **`spec_info.SpecInput`** — `@runtime_checkable` Protocol exposing:
  - `is_draft_input()` / `is_verify_input()`
  - `get_logical_token_num()` / `get_allocated_token_num()` / `get_verify_token_num()` — the three token counts the RFC requires to be separated (scheduler advance / KV trimming / verify-metadata+DP-accounting respectively)
  - `get_spec_adjust_token_coefficient()` / `filter_batch()` / `merge_batch()`
  - Docstring fixes the host/device boundary rules and DP-padded layout (Route 1: target+draft both DP) semantics
- **`EagleDraftInput`** — implements `SpecInput`; per-field docstrings annotate device vs host, shape, JIT-cache-key participation, and the `allocate_lens` ≠ `accept_length` ≠ `new_seq_lens` distinction. Multi-layer MTP cross-round hidden_states semantics noted.
- **`EagleVerifyInput`** — implements `SpecInput`; per-field docstrings annotate device vs host and static-metadata fields; `filter_batch`/`merge_batch` raise (verify input is single-round).
- **`test_spec_info.py`** — Protocol conformance + three-token-count distinction.

## Non-changes

- No behavior change to the existing dp=1 path (pure additions: docstrings, Protocol, new methods).
- `GenerationBatchResult` already has all RFC-required fields except optional `new_seq_lens` (derivable from `seq_lens + accept_lens`); not adding it here.
- Type annotations on device fields changed `np.ndarray` → `jax.Array | None` to reflect actual runtime types after #1063's reshard changes — annotations only, no code path change.

## Open questions for P1-2 author

1. Should `SpecInput` be a `Protocol` (current, duck-typed) or an `ABC` base class? Protocol avoids inheritance coupling but `ABC` lets `BaseSpecWorker` type-hint more strictly.
2. `get_logical_token_num` currently returns a scalar `int` (sum across batch). Should it return per-request `np.ndarray` instead for DP per-rank accounting?
3. `EagleDraftInput.merge_batch` currently uses `np.concatenate` (host-side). RFC says these are device arrays — should merge happen on device (`jnp.concatenate`)?

## Test plan

- [x] `test_spec_info.py` 3/3 passed (CPU)
- [x] pre-commit clean
- [ ] dp=1 EAGLE E2E unchanged (will verify after rebasing onto merged #1063)

### CODE REVIEW

File: python/sgl_jax/srt/speculative/spec_info.py [hunk 1]
```diff
@@ -1,13 +1,57 @@
...
 import jax
+import numpy as np
 
 from sgl_jax.srt.layers.logits_processor import LogitsProcessorOutput
 
 logger = logging.getLogger(__name__)
 
 
+@runtime_checkable
+class SpecInput(Protocol):
```
Comment (pengchengneo) [inline python/sgl_jax/srt/speculative/spec_info.py:16]: I’d keep `SpecInput` as a Protocol for now. In this repo it is mainly a contract `ModelWorkerBatch.spec_info`, not shared implementation, and the current inputs are dataclass+pytree objects. An ABC would add inheritance coupling before P1-2 proves we need shared base behavior.

We can make the Protocol strict with clear return semantics and boundary checks, then add an ABC/mixin later if real shared logic appears.

File: python/sgl_jax/srt/speculative/eagle_util.py [hunk 2]
```diff
@@ -687,22 +720,65 @@ class EagleVerifyOutput:
...
+    def get_allocated_token_num(self) -> np.ndarray | None:
+        return None
+
+    def get_verify_token_num(self, bs: int) -> int:
+        return bs * self.draft_token_num
+
+    def filter_batch(self, new_indices: np.ndarray, has_been_filtered: bool = True) -> None:
+        raise NotImplementedError("EagleVerifyInput is consumed within one round")
+
+    def merge_batch(self, other) -> None:
```
Comment (pengchengneo) [inline python/sgl_jax/srt/speculative/eagle_util.py:779]: Although `merge_batch` is called from the scheduler host path, these fields are next-round draft state and are consumed on device.

You can replace this with jnp.concatenate

and in phase3 we will put these field on FutureMap, just concatenate future indices will be ok

File: python/sgl_jax/srt/speculative/spec_info.py [hunk 3]
```diff
@@ -1,13 +1,57 @@
...
+    """
+
+    def is_draft_input(self) -> bool: ...
+    def is_verify_input(self) -> bool: ...
+
+    def get_spec_adjust_token_coefficient(self) -> int:
+        """Multiplier for scheduler new-token budgeting (e.g. draft_token_num)."""
+        ...
+
+    def get_logical_token_num(self, bs: int) -> int: ...
```
Comment (pengchengneo) [inline python/sgl_jax/srt/speculative/spec_info.py]: Returning only a batch-level int loses DP rank structure, this function can return per-request counts, e.g. `np.ndarray`, and let callers sum when they need a scalar.

File: python/sgl_jax/srt/speculative/spec_info.py (hunk 1 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/spec_info.py:16]: Kept as Protocol. Will revisit if P1-2 surfaces shared base behavior worth an ABC/mixin.

File: python/sgl_jax/srt/speculative/eagle_util.py (hunk 2 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/eagle_util.py:779]: Switched the four device fields (`hidden_states`/`verified_id`/`topk_p`/`topk_index`) to `jnp.concatenate`; left `allocate_lens` on `np.concatenate` (host array per the contract). 69113de5.

File: python/sgl_jax/srt/speculative/spec_info.py (hunk 3 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/spec_info.py]: Changed to per-request `np.ndarray` (host int32 `(bs,)`); callers sum for batch totals. Updated both implementations + the Protocol docstring + the unit test. 69113de5.

**Review by pengchengneo — APPROVED** [review]

**Comment by pengchengneo** [comment]
LGTM
